### ISSUE TITLE
approachable docs for the documentation hosting

### ISSUE THREAD

**Issue by RonnyPfannschmidt** [devpi/devpi#770]
while trying to  explain devpis documentation hosting feature to a coworker, i noted, that while the capability s indeed there, the details about it are mostly technical bits all over the place

some examples of how to upload docs and how to integrate intersphinx in a single place would be a great help to start quickly

**Comment by exhuma** [comment]
@RonnyPfannschmidt I'm currently trying to get it integrated with intersphinx and can't figure out how. It sounds like you managed to do it. How did you?

**Comment by RonnyPfannschmidt** [comment]
I didn't, and we no longer use devpi for that part of the infrastructure

**Comment by exhuma** [comment]
With "that part of the infrastructure" do you mean docs hosting only? Or the internal package repo?

We're currently running devpi in a transitional mode (moving *towards* devpi). Your comment worries me. We chose devpi as I didn't find any alternative and there are things I find incovenient, but manageable with it.

**Comment by fschulze** [comment]
Please let us know about what you find inconvenient so we can address it if possible. I plan to spend some time on reworking the docs this month.

**Comment by exhuma** [comment]
This may sound "ranty", but it's not intended to be (really)...

The main point is that devpi just does things "differently" than you might expect. There's no simple configuration via a config-file (it has to be done via the devpi client) and the files are stored with a weird directory structure, even including plusses which already caused some issues (I don't remember the details).

I assume, they are all design-decisions, but they make devpi this "odd one out" in all our DevOps scripts.

I also think to remember that it has some stuff stored inside a sqlite file (or some other embedded format) which is hard to get to once its running in a container. I'd rather have DB related stuff in our main postgres cluster. It would then benefit from all our monitoring and maintenance.

The way I see it, changing these aspects of devpi would likely be difficult and I don't feel comfortable asking these changes as they are wide-ranging. So I rather see devpi for us a not being the "right tool for the job". Unfortunately I'm not aware of any other simple package repo which is compatible with twine. So we make due and accept the inconveniences.

A very very important thing to note though is that we are a fairly small team with too much work on our hand and, maybe we just use devpi incorrectly or misunderstand some concepts. So maybe a docs rework might actually shed some new light on everything.

All we really want is to be able to `twine upload` somewhere and then `pip install` from there too (with a fallback to pypi).

**Comment by fschulze** [comment]
I consider devpi-server very database like. Everything you do with devpi-client to configure devpi is in my opinion the same you do with a database client. You don't define database schemas with configuration files. Though I have to admit, that you can define schemas in a text file via sql, but I don't consider that a config file, because changes aren't reflected in the db. And you can do full export/imports of all the data, but not for specific indexes like for tables in a db.

Everything you configure via the command line can be set in a configuration file and used with the ``-c`` option or if the config file is in a standard location. https://devpi.net/docs/devpi/devpi/stable/+doc/quickstart-server.html?highlight=configuration#using-a-configuration-file-for-devpi-server

You can use PostgreSQL with devpi: https://pypi.org/project/devpi-postgresql/
(Almost) Everything will then be in PostgreSQL, including the release files. There are still some open ends there:
https://github.com/devpi/devpi/milestone/1
If you don't need the search in the web UI, then you can use ``--indexer-backend=null`` to avoid the files for that.
That only leaves the ``.nodeinfo`` in the server directory. I hope to get rid of it as well in the not too distant future.

**Comment by smartsammler** [comment]
> @RonnyPfannschmidt I'm currently trying to get it integrated with intersphinx and can't figure out how. It sounds like you managed to do it. How did you?

@hoefling maybe has managed to use intersphinx together with devpi. At least I read his setup in https://github.com/sphinx-doc/sphinx/issues/3095 like it.

But the URL in the error logs looks like it was using intersphinx with a pinned version of the documentation.

@fschulze 
If I remember correctly when I tried to figure out the URL to use for a button to the latest docs
(https://devpi.server/user_name/prod/project_name/stable/+d/index.html) I found it quite hard to find this information.
Maybe one can include this information at the same point as the interspinx-URL.

Thank you for devpi and devpi-web anyway.

**Comment by hoefling** [comment]
TL;DR the URL to put into intersphinx config is `http://<host>:<port>/<user>/<index>/<pkgname>/<pkgversion>/+doc/`, nothing else is needed. Proper configuration example below.

My 2c first: the only inconvenience I found using `devpi` is that the docs upload command relies on the presence of the `setup.py` script, so one has to ensure the `python setup.py build_sphinx` is runnable. As we switched our projects to `poetry` and PEP 517, I had to resort to zip the docs and upload the zip file which doesn't feel quite right, but works. A snippet from the CI configuration:
```sh
- poetry run sphinx-build -EWanb html doc/ build/
- poetry run python -m zipfile -c dist/${PKGNAME}-${PKGVER}.doc.zip build/*
- poetry run devpi use ${DEVPI_INDEX_URL}
- poetry run devpi login ${DEVPI_USER} --password=${DEVPI_PASSWORD}
- poetry run devpi upload --only-docs dist/*.doc.zip --index ${DEVPI_USER}/staging
```

As for intersphinx:

The only task here is to find the link to the objects inventory file and use it for crossreferencing configuration. At least when building HTML docs, Sphinx will generate the inventory file in the build target directory (named `objects.inv`). When you zip and upload the docs to `devpi`, it will be hosted under
```http://<host>:<port>/<user>/<index>/<pkgname>/<version>/+doc/objects.inv```
Example: we have `devpi` running behind `https://pypi.devnet.netqa.de`, the user is `netqa`, index is `staging`, package is `publiqa-core`. The link we use is thus `https://pypi.devnet.netqa.de/netqa/staging/publiqa-core/latest/+doc/objects.inv`, as we always reference to the `latest` version on staging and to the `stable` version on prod.

Verify the `objects.inv` is accessible. This should list all the objects in the documentation you can crossref with:
```sh
$ python -m sphinx.ext.intersphinx https://pypi.devnet.netqa.de/netqa/staging/publiqa-core/latest/+doc/objects.inv
py:attribute
	foo.bar.CONSTANT       api/foo.html#foo.bar.CONSTANT
py:class
	spam.eggs.Bacon        api/spam_api.html#spam.eggs.Bacon
	eggs.BaconWithEggs     api/eggs_api.html#eggs.BaconWithEggs
py:function
	foo.bar.baz            api/foo.html#foo.bar.baz
```
Now configure the intersphinx crossref. The configuration is as defined in [intersphinx docs](https://www.sphinx-doc.org/en/master/usage/extensions/intersphinx.html):
```py
intersphinx_mapping = {
    "unique_key": ("parent_of_objects_inv_url", "objects_inv_file_name_or_None")
}
```
Example: if another project `foo` should reference `publiqa-core` in its docs, the configuration in `foo`'s `conf.py` will look like this:
```py
intersphinx_mapping = {
    "python": ("https://docs.python.org/3/", None),
    ...
    "publiqa": (
        "https://pypi.devnet.netqa.de/netqa/staging/publiqa-core/latest/+doc/",
        None,
    ),
}
```
This means that Sphinx will download `https://pypi.devnet.netqa.de/netqa/staging/publiqa-core/latest/+doc/objects.inv` when invoking HTML builder and use its objects when resolving crossrefs. `None` indicates that the inventory file is named `objects.inv`, no special naming here. The unique key `publiqa` is to avoid unlikely ambiguities: for example, if the project would define its own `zipfile.ZipFile`, the reference ``:class:`zipfile.ZipFile` `` would lead to Python docs instead. To avoid this, one would prefix the key: `` :publiqa:class:`zipfile.ZipFile` ``.

**Comment by hoefling** [comment]
So basically, the crossref'ing works like a charm once you know what URL to put into `intersphinx_mapping`. I even remember from the past days of using the `setup.py` that the long deprecated `python setup.py upload_docs` worked with `devpi`; however, there were some issues with that on Sphinx side so I had to subclass `build_sphinx` and fix some methods, so I wouldn't recommend using it anyway.

**Comment by fschulze** [comment]
@hoefling 
> My 2c first: the only inconvenience I found using devpi is that the docs upload command relies on the presence of the setup.py script, so one has to ensure the python setup.py build_sphinx is runnable.

Can you propose an alternative you would like to see?

> As we switched our projects to poetry and PEP 517, I had to resort to zipping the docs and uploading the zip file which doesn't feel quite right, but works.

In what way doesn't this feel right? The documentation *is* just a zip file.

**Comment by hoefling** [comment]
> Can you propose an alternative you would like to see?
I agree this isn't doable for the docs build as there isn't any spec for that at all. However, at least for building dists `devpi` should leverage build backends according to specs in PEP 517/518, most probably querying the build backends from `pyproject.toml` and calling `backend.build_wheel` or `backend.build_sdist`. The `setup.py` remains as the legacy option. In my case, `poetry publish` was an alternative, but this requires to configure the whole host/credentials/certs machinery also for `poetry`; I wouldn't mind to continue using `devpi upload` instead.

As for the docs, I would imagine some config where one can specify the backend like
```yaml
sphinx:
  builder: html
  configuration: doc/conf.py
```
or
```yaml
mkdocs:
  configuration: doc/mkdocs.yml
```
and `devpi upload --with-docs` invokes the `sphinx-build -b html -c doc/conf.py` or `mkdocs build --config-file doc/mkdocs.yml` and zip/upload the resulting artifact under the hood. Kind of RTD-style.

> In what way doesn't this feel right? The documentation _is_ just a zip file.
My primary concern is that this isn't specified anywhere, so if e.g. `devpi` decides to switch to tars for some reason, my builds will break. A command that manages building and uploading the artifacts is IMO the optimal way, without me having to worry about the intermediate steps.

The good news is that right now, one can upload any zip file with an `index.html` in it - which is great for hosting stuff like `mkdocs` documentation, I also use `devpi` for hosting project specification documents, translated to HTML.
